### PR TITLE
Zenodo

### PR DESCRIPTION
**Pull request by dham**
Zenodo integration for Firedrake:

* firedrake-zenodo script which allows users and developers to easily create releases.

* Migration of bitbucket repositories to github.

* Documentation on the website and in the script.

### CODE REVIEW

**Comment by miklos1** [comment]
I found that not just PETSc, but also FEniCS has an official GitHub mirror: https://github.com/FEniCS

**Comment by miklos1** [comment]
I see this also contains #633.

**Comment by dham** [comment]
Yes, it seemed silly to write in a whole load of new logging the old way.

File: scripts/firedrake-install [hunk 1]
```diff
@@ -301,6 +316,17 @@ def git_update(name):
...
     else:
+        # Ensure remotes get updated if and when we move repositories.
+        if url:
+            plain_url = split_requirements_url(url)[1]
+            current_url = check_output(["git", "remote", "-v"]).split()[1]
+            protocol = "https" if current_url.startswith("https") else "ssh"
+            new_url = git_url(plain_url, protocol)
+            # Ensure we only change from bitbucket to github and not the reverse.
+            if new_url != current_url and "bitbucket.org" in current_url \
+               and "github.com/firedrakeproject" in plain_url:
```
Comment (wence-) [inline scripts/firedrake-install]: So this works because plain_url is always `/` separated (because it comes from the requirements file).

File: scripts/firedrake-install [hunk 2]
```diff
@@ -301,6 +316,17 @@ def git_update(name):
         else:
             check_call(["git", "checkout", travis_commit])
     else:
+        # Ensure remotes get updated if and when we move repositories.
+        if url:
+            plain_url = split_requirements_url(url)[1]
+            current_url = check_output(["git", "remote", "-v"]).split()[1]
```
Comment (wence-) [inline scripts/firedrake-install]: This relies on `origin` being the only remote and someone not configuring a separate push and fetch urls.  I guess this is probably a sane default assumption.  If you wanted to separately find the push and fetch urls, guaranteeing that you did so for `origin`, you would do:

```
push_url = check_output(["git", "remote", "get-url", "--push", "origin"])
fetch_url = check_output(["git", "remote", "get-url", "origin"])
```

File: scripts/firedrake-zenodo [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            log.error("Failed to retrieve git hash for %s" % component)
+            raise
+    return shas
+
+if args.release_tag:
+    tag = args.release_tag[0]
+    log.info("Retrieving BibTeX data for Firedrake release %s." % tag)
+    log.info("This may take a few seconds.")
+    try:
+        request = urlopen("http://firedrakeproject.org/zenodo-bibtex.cgi?tag=%s" % tag)
```
Comment (wence-) [inline scripts/firedrake-zenodo]: This makes a web request to `firedrakeproject.org/zenodo-bibtex.cgi` passing a user-specified tag.  It is not sanitised here, so let's check that the cgi script does sanitising.

File: scripts/zenodo-bibtex.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        parser = BibtexFinder()
+        parser.feed(html.read())
+        if parser.bibtexdata:
+            bibtex.append(parser.bibtexdata)
+
+    return "\n\n".join(bibtex)
+
+
+form = cgi.FieldStorage()
+firedrake_tag = form.getfirst("tag")
```
Comment (wence-) [inline scripts/zenodo-bibtex.py]: OK, so we get the tag that the user provided.

File: scripts/zenodo-bibtex.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+if not firedrake_tag:
+    fail("400 Bad Request", "Request must specify the tag.")
```
Comment (wence-) [inline scripts/zenodo-bibtex.py]: If it's empty, we fail.

File: scripts/zenodo-bibtex.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+# Do some primitive parsing to avoid more blatant security issues.
+try:
+    assert firedrake_tag.startswith("Firedrake_")
```
Comment (wence-) [inline scripts/zenodo-bibtex.py]: It's not empty.  It has to start with `"Firedrake_"`

File: scripts/zenodo-bibtex.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    float(firedrake_tag[10:])
```
Comment (wence-) [inline scripts/zenodo-bibtex.py]: I need to be able to cast the remainder of the tag to a `float`.  We pass the entire rest of the user-provided string to `float`, so a malformed input *could* expose a bug in core python.  We know something more about the tag, right.  If it's a valid tag, it must be `YYYYMMDD.NNNN`. So we could do:

```python
date = int(firedrake_tag[10:18])
assert tag[18] == "."
# assume we will not create more than 9999 tags on the same commit
number = tag[19:23]
```

File: scripts/zenodo-bibtex.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+except:
+    fail("400 Bad Request", "%s is not a legal Firedrake release tag" % firedrake_tag)
```
Comment (wence-) [inline scripts/zenodo-bibtex.py]: A bad tag is echoed back in full to the user.  I'd be tempted just to say "Provided tag is not a legal Firedrake release tag, it should have the format `Firedrake_YYYYMMDD.NNNN`..."?

File: scripts/zenodo-bibtex.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+def find_ids(response, firedrake_tag):
+
+    ids = []
+    for record in response:
+        try:
+            for id in record['metadata']['related_identifiers']:
+                if id['identifier'].endswith(firedrake_tag):
+                    ids.append(record['record_id'])
+        except KeyError:
```
Comment (wence-) [inline scripts/zenodo-bibtex.py]: So you spin over the json blobs provided by zenodo listing your depositions and collect those that match the provided tag.

File: docs/source/zenodo.rst [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+These DOIs can be used in citations in publications to provide a
+reference to the exact version of the software used, and thereby to
+improve the reproducibility of your computational science.
+
+How to register DOIs for a version of Firedrake
+-----------------------------------------------
+
+This section assumes that you have a Firedrake installation which you
+have used to conduct some numerical experiment and which you wish to
+publish or otherwise record for posterity. It is assumed that your
```
Comment (wence-) [inline docs/source/zenodo.rst]: I found this sentence a little hard to parse.  Unfortunately, in attempting to come up with something better I thus far have nothing.

File: docs/source/zenodo.rst [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+What else do you need to do?
+----------------------------
+
+Archive your code
+~~~~~~~~~~~~~~~~~
+
+``firedrake-zenodo`` produces citable DOIs which point to the versions
+of Firedrake components you used. This covers your bases as far as
+Firedrake concerns, but doesn't cover your code which uses
```
Comment (wence-) [inline docs/source/zenodo.rst]: "as far as concerns Firedrake" or "as far as Firedrake is concerned"?

File: docs/source/zenodo.rst [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+Cite your sources
+~~~~~~~~~~~~~~~~~
+
+Citing custom DOIs for particular versions of Firedrake and its
+dependencies aids readers of your papers in reproducing your
+science. However it's a supplement to, and not a replacement for,
+citing the published resources for the computational methods you are
+employing. Firedrake also offers support for citing the papers on
+which your computations depend. This is documented on the
+:doc:`publications` page.
```
Comment (wence-) [inline docs/source/zenodo.rst]: Not in this change, but do we want to split out the "how to cite firedrake" from the publications page into a separate page?

**Comment by wence-** [comment]
Makefile needs to be updated to lint the new scripts.

**Comment by miklos1** [comment]
So GitHub FIAT isn't up-to-date with Bitbucket FIAT, and I can't fix that because of repository permissions.

**Comment by wence-** [comment]
I've rebased (so that the branch should now pass tests) and made `make lint` lint all the scripts, not just firedrake-install.  Hopefully this will go green and then we can merge.

**Comment by wence-** [comment]
Good, merging.

File: scripts/firedrake-zenodo [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    out = file(cwd+"/"+args.output_file[0], "w")
+    out.write(json.dumps(shas) + "\n")
+
+    log.info("Wrote release information to %s" % args.output_file[0])
+    sys.exit(0)
+
+try:
+    import github3
+except ImportError:
+    "Publishing releases requires the github3 module. Please pip install github3.py"
```
Comment (kynan) [inline scripts/firedrake-zenodo]: Don't you want to `sys.exit(1)` after this?
